compiler/luci/pass/src/FuseRoPEPass.cpp

```diff
@@ -0,0 +1,281 @@
+{
+  CHECK_OR_FALSE(luci::fill(&mul_cos, &mul_sin).with_commutative_args_of(add_ofm));
+  CHECK_OR_FALSE(luci::fill(&ifm, &cos_table).with_commutative_args_of(mul_cos));
+
+  auto ifm_circle = loco::must_cast<luci::CircleNode *>(ifm);
+  CHECK_OR_FALSE(ifm_circle->shape_status() == luci::ShapeStatus::VALID);
+  CHECK_OR_FALSE(ifm_circle->rank() == 4);
+  CHECK_OR_FALSE(ifm_circle->dim(3).known());
+
+  CHECK_OR_FALSE(luci::fill(&concat, &sin_table).with_commutative_args_of(mul_sin));
+  CHECK_OR_FALSE(concat->numValues() == 2);
+
+  strided_slice = dynamic_cast<luci::CircleStridedSlice *>(concat->values(1));
+  CHECK_OR_FALSE(strided_slice);
+
+  neg = dynamic_cast<luci::CircleNeg *>(concat->values(0));
+  CHECK_OR_FALSE(neg);
+
+  strided_slice_neg = dynamic_cast<luci::CircleStridedSlice *>(neg->x());
+  CHECK_OR_FALSE(strided_slice_neg);
+
+  _matched = true;
+  return true;
+}
+
+bool RoPEPattern::matched()
+{
+  if (_matched)
+    return true;
+
+  switch (_pm)
+  {
+    case PatternMode::Mode_NEOX:
+      return match<PatternMode::Mode_NEOX>();
+
+    default:
+      break;
+  }
+
+  throw std::runtime_error("Invalid RoPE Pattern.");
+}
+
+#undef CHECK_OR_FALSE
+
+class FuseRoPE final
+{
+public:
+  FuseRoPE(const RoPEPattern &p) : _p(p) {}
+
+public:
+  void apply(void);
+
+private:
+  template <RoPEPattern::PatternMode> void apply(void);
+
+private:
+  luci::CircleRoPE *create_rope(loco::Graph *graph);
+
+private:
+  const RoPEPattern &_p;
+};
+
+luci::CircleRoPE *FuseRoPE::create_rope(loco::Graph *graph)
+{
+  assert(graph);
+
+  auto rope = graph->nodes()->create<luci::CircleRoPE>();
+  rope->input(_p.ifm);
+  rope->cos_table(_p.cos_table);
+  rope->sin_table(_p.sin_table);
+
+  if (_p.mode() == RoPEPattern::PatternMode::Mode_NEOX)
+    rope->mode(luci::RoPEMode::GPT_NEOX);
+
+  rope->name(_p.add_ofm->name() + "_rope");
+  return rope;
+}
+
+template <> void FuseRoPE::apply<RoPEPattern::PatternMode::Mode_NEOX>()
+{
+  auto graph = _p.add_ofm->graph();
+
+  auto rope = create_rope(graph);
+
+  // set origin
+  std::vector<std::shared_ptr<luci::CircleNodeOrigin>> origin_vec{
+    luci::get_origin(_p.strided_slice), luci::get_origin(_p.strided_slice_neg),
+    luci::get_origin(_p.neg),           luci::get_origin(_p.concat),
+    luci::get_origin(_p.mul_cos),       luci::get_origin(_p.mul_sin),
+    luci::get_origin(_p.add_ofm)};
+
+  luci::add_origin(rope, luci::composite_origin(origin_vec));
+
+  rope->cos_table(_p.cos_table);
+  rope->sin_table(_p.sin_table);
+
+  replace(_p.add_ofm).with(rope);
+}
+
+void FuseRoPE::apply()
+{
+  assert(_p.matched());
+
+  switch (_p.mode())
+  {
+    case RoPEPattern::PatternMode::Mode_NEOX:
+      apply<RoPEPattern::PatternMode::Mode_NEOX>();
+      break;
+
+    default:
+      break;
+  }
+}
+} // namespace
+
+namespace
+{
+
+bool fuse_rope(luci::CircleAdd *add)
+{
+  RoPEPattern::PatternMode pm = RoPEPattern::PatternMode::Mode_NEOX;
+
+  // TODO support other modes (ex. GPT_J)
```

Comment:
Currently, only GPT-NEOX is supported, and the default is GPT-NEOX.
If another mode (ex. GPT-J) is added in the future, need to find the pattern and add the code to distinguish the mode.
so this has not been confirmed yet and will be added in the future